File: strategy/rsi.go
package strategy

import (
	"trader/events"
	"trader/types"

	"github.com/markcheno/go-talib"
	"github.com/rs/zerolog/log"
)

var Period = 14

type RsiConfig struct {
	Overbought float64
	Oversold   float64
}
type Rsi struct {
	id     string
	config RsiConfig
	pubsub events.PubSub
	states map[string]*State
}

type State struct {
	closes  []float64
	holding bool
}

// Init state map for each symbol
func makeState(symbols []string) map[string]*State {
	states := make(map[string]*State)

	for _, symbol := range symbols {
		closes := []float64{}
		holding := false

		state := State{closes, holding}
		states[symbol] = &state
	}

	return states
}

func NewRsi(id string, config RsiConfig, pubsub events.PubSub, symbols []string) Rsi {
	log.Trace().
		Str("ID", id).
		Float64("overbought", config.Overbought).Float64("oversold", config.Oversold).
		Msg("Strategy.Rsi.Init")

	states := makeState(symbols)

	return Rsi{id, config, pubsub, states}
}

func (r *Rsi) Predict(k types.Kline, symbol string) {
	// Return if Kline isn't closed yet
	if !k.Closed {
		return
	}

	state := r.states[symbol]

	state.closes = append(state.closes, k.Price)

	log.Info().Str("symbol", symbol).Float64("price", k.Price).Bool("closed", k.Closed).Msg(r.id)

	if len(state.closes) > Period {
		rsi := talib.Rsi(state.closes, Period)
		last := rsi[len(rsi)-1]

		log.Debug().
			Str("symbol", symbol).
			Float64("price", k.Price).
			Bool("holding", state.holding).
			Float64("last_rsi", last).
			Msg(r.id)

		if last > r.config.Overbought {
			if state.holding {
				r.pubsub.Publish(events.SignalSell, events.SignalSellPayload{symbol})
			} else {
				log.Warn().Str("symbol", symbol).Float64("last_rsi", last).Msg("Rsi.Overbought.NoPosition")
			}
		}

		if last < r.config.Oversold {
			if state.holding {
				log.Warn().Str("symbol", symbol).Float64("last_rsi", last).Msg("Rsi.Oversold.InPosition")
			} else {
				r.pubsub.Publish(events.SignalBuy, events.SignalBuyPayload{symbol})
				state.holding = true
			}
		}
	}
}

File: services/strategy/internal/rsi.py
import talib
import numpy
from utils.pubsub import PubSub
from utils.events import Events
from utils.log import Logger

PERIOD = 14
OVERBOUGHT = 70
OVERSOLD = 30


class RSI:
    pubsub = None

    holding = False
    closes = []

    log = Logger.get('RSI')

    def __init__(self, pubsub):
        self.log.info(
            f'Strategy.Rsi.Init overbought={OVERBOUGHT} oversold={OVERSOLD}')
        self.pubsub = pubsub

    async def predict(self, data):
        symbol = data['Symbol']
        kline = data['Kline']

        closed = kline['Closed']
        price = float(kline['Price'])

        if not closed:
            return

        self.closes.append(price)

        total_closes = len(self.closes)

        if total_closes > PERIOD:
            np_closes = numpy.array(self.closes)
            rsi = talib.RSI(np_closes, PERIOD)
            last_rsi = rsi[-1]

            self.log.debug(
                f'Rsi symbol={symbol} price={price} last_rsi={last_rsi}')

            if last_rsi > OVERBOUGHT:
                if self.holding:
                    payload = {
                        'Side': 'SELL',
                        'Symbol': symbol,
                        'Price': price
                    }
                    self.pubsub.publish(Events.SignalTrade, payload)
                    self.holding = False  # TODO: Remove hold if order succeeds
                else:
                    self.log.warn('Rsi.Overbought.NoPosition')

            if last_rsi < OVERSOLD:
                if self.holding:
                    self.log.warn('Rsi.Oversold.InPosition')
                else:
                    payload = {
                        'Side': 'BUY',
                        'Symbol': symbol,
                        'Price': price
                    }
                    self.pubsub.publish(Events.SignalTrade, payload)
                    self.holding = True  # TODO: trigger hold only if order succeeds
